wgsl/index.bs

```diff
@@ -5690,6 +5690,370 @@ Note: If the function [=return value|returns a value=], that value is ignored.
+As a consequence of the rules above, a function behavior is always one of {Next}, {Discard} or {Next, Discard}.
+
+Similarly, we assign a [=behavior=] to each expression, since expressions can include function calls, which can discard.
+Like functions, expression behaviors are always one of {Next}, {Discard}, or {Next, Discard}.
+
+Note: There is currently no valid program with an expression that has a [=behavior=] of {Discard}.
+The reason is that only functions without a return type can have such a [=behavior=], and there is no compound expression in which such a function can be called.
+
+<table class='data'>
+  <caption>Rules for analyzing and validating the behaviors of statements</caption>
+  <thead>
+    <tr><th>Statement<th>Preconditions<th>Resulting behavior
+  </thead>
+  <tr>
+    <td class="nowrap">*s1* ; *s2*
+    <td class="nowrap">*s1*: *B1*<br>
+        Next in *B1*<br>
+        *s2*: *B2*
+    <td class="nowrap">*B1*&#x2216;{Next} &cup; *B2*
+  <tr>
+    <td class="nowrap">var x:T;
+    <td>
+    <td>{Next}
+  <tr>
+    <td class="nowrap">let x = *e*;
+    <td class="nowrap">*e*: *B*
+    <td>*B*
+  <tr>
+    <td class="nowrap">var x = *e*;
+    <td class="nowrap">*e*: *B*
+    <td>*B*
+  <tr>
+    <td class="nowrap">x = *e*;
+    <td class="nowrap">*e*: *B*
+    <td>*B*
+  <tr>
+    <td class="nowrap">f(*e_1*, ..., *e_n*);
+    <td class="nowrap">*e_1*: *B_1*<br>
+        ...<br>
+        *e_n*: *B_n*<br>
+        *f*'s body has behavior *B*
+    <td class="nowrap">*B* &cup; (*B_1* &cup; ... &cup; *B_n*)&#x2216;{Next}
+  <tr>
+    <td>return;
+    <td>
+    <td>{Return}
+  <tr>
+    <td class="nowrap">return *e*;
+    <td class="nowrap">*e*: *B*<br>
+    <td class="nowrap">*B*&#x2216;{Next} &cup; {Return}
+  <tr>
+    <td>discard;
+    <td>
+    <td>{Discard}
+  <tr>
+    <td>break;
+    <td>
+    <td>{Break}
+  <tr>
+    <td>continue;
+    <td>
+    <td>{Continue}
+  <tr>
+    <td>fallthrough;
+    <td>
+    <td>{Fallthrough}
+  <tr>
+    <td class="nowrap">if (*e*) *s1* elseif *s2* ...  else *sn*
+    <td class="nowrap">*e: B*<br>
+        *s1*: *B1*<br>
+        ...<br>
+        *sn*: *Bn*
+    <td class="nowrap">*B*&#x2216;{Next} &cup; B1 &cup; ... &cup; Bn
+  <tr>
+    <td class="nowrap">loop {*s*}
+    <td class="nowrap">*s*: *B*
+    <td class="nowrap">let b = *B*&#x2216;{Next};<br>
+        if (Break in b) b = b&#x2216;{Break} &cup; {Next};<br>
+        b&#x2216;{Continue}
+  <tr>
+    <td class="nowrap">loop {*s1* continuing {*s2*}}
+    <td class="nowrap">*s1*: *B1*<br>
+        *s2*: *B2*<br>
+        Neither Continue nor Return are in *B2*
+    <td class="nowrap">let b = (*B1* &cup; *B2*)&#x2216;{Next};<br>
+        if (Break in b) b = b&#x2216;{Break} &cup; {Next};<br>
+        b&#x2216;{Continue}
+  <tr>
+    <td class="nowrap">switch(*e*) {case _: *s1* ... case _: *sn*}
+    <td class="nowrap">*e*: *B*<br>
+        *s1*: *B1*<br>
+        ...<br>
+        *sn*: *Bn*<br>
```

Comment:
nitpick: Please express that there can be a default clause.

Reply:
Good idea.
And very good catch regarding infinite loops, I'll remove the note I had added saying that empty behaviors cannot happen.